First version of the profile publisher.

fiji-flo opened the pull request:
Squashed and signed.

fiji-flo committed:
First version of the profile publisher.

Squashed and signed.

johngian commented on .dockerignore:
```diff
@@ -0,0 +1,3 @@
+.git/
+**/node_modules/
```
why isn't ``node_modules/`` enough?

fiji-flo replied:
it is, will change this.

johngian commented on .dockerignore:
```diff
@@ -0,0 +1,3 @@
+.git/
+**/node_modules/
+test/
```
I don't see the value of not including part of the codebase to the docker image. Is there any reasoning around it?

fiji-flo replied:
I'm used to just ship production _binaries_. The dev dependencies are missing, too. But that's not a strong opinion.
@johngian What's the value of including them?

johngian commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+**/node_modules
```
same here

johngian commented on .travis.yml:
```diff
@@ -0,0 +1,7 @@
+install:
+  - npm install
+script:
+  - npm test
```
Is the coverage part handled automatically in the test command ?

fiji-flo replied:
No. I postponed the coverage part. Ideally I wanna use _codecov.io_

johngian commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
+COPY package.json .
+RUN npm install --production
+COPY . .
```
nit: I think its more readable to do ``COPY . /app`` but feel free to leave it as it is.

johngian commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+### User Updates
+
+`/userupdate` is an endoint used by mozillinas to publish profil updates to CIS.
```
typo: ``mozillians``, ``profile updates``

johngian commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+Posting a profile update to this endpoint will trigger the following flow:
+
+1. Profile gets posited to the CIS endpoint, which will return an `updateId`
```
typo: ``posited``

johngian commented on README.md:
```diff
@@ -0,0 +1,65 @@
+`/externalupdate` is an endpoint used by CIS to trigger profile updates for the
+DinoPark services.
+
+Posting a an update event to this endpoint will trigger the following flow:
```
typo: ``a an`` --> ``an``

johngian commented on config.json:
```diff
@@ -0,0 +1,9 @@
+{
```
Can we setup a more 12factor friendly config management?

johngian commented on index.js:
```diff
@@ -0,0 +1,36 @@
+
+    app = new App(cfg);
+
+    logger.info(`starting to serve on port: ${cfg.port}`);
```
Let's be consistent in logging: ``starting`` --> ``Starting``

johngian commented on index.js:
```diff
@@ -0,0 +1,36 @@
+    logger.info("shut down gracefully :)");
+    process.exit(0);
+  } catch (e) {
+    logger.error(`faild to stop: ${e}`);
```
``faild``

johngian commented on lib/config.js:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+const VALID = [
```
instead of running our own validation can't we just use a lib to handle config? This is also relevant to the comment above regarding 12 factor.

johngian replied:
also instead of creating a second copy of the ``VALID`` values (which is pretty prone to errors) can we annotate the config for required fields and expected values ?

johngian commented on lib/externalUpdate.js:
```diff
@@ -0,0 +1,51 @@
+        })
+        .catch(e => {
+          logger.error(e);
+          res.status(503);
```
Should we add a more informative response error message here instead of status code?

johngian commented:
nit: In logging let's be consistent on how long messages start (either upper or lower case, not mixed)

johngian commented on lib/externalUpdate.js:
```diff
@@ -0,0 +1,51 @@
+
+  async getAndDistributeUpdate(userId) {
+    return this.personApi(userId)
+      .then(updatedProfile => this.searchService(updatedProfile))
```
Should we keep this list as a config somewhere or the flow is only going to be triggered here.

fiji-flo replied:
I'd keep it here for now. I don't see that this will change any time soon. And it's easy to factor this out in the future.

johngian commented on lib/personApi.js:
```diff
@@ -0,0 +1,18 @@
+      if (res.status === 200 && res.data) {
+        return Promise.resolve(res.data);
+      } else {
+        return Promise.reject(`PersonApi return status code ${res.status}`);
```
What happens if ``res.status === 200`` but no data exist?

johngian replied:
I think the error message is going to be kinda misleading

johngian commented on lib/updateOrgchart.js:
```diff
@@ -0,0 +1,22 @@
+        return Promise.resolve(profileUpdate);
+      } else {
+        return Promise.reject(
+          `orgchart service return status code ${res.status}`
```
Same here.

johngian commented on lib/updateSearch.js:
```diff
@@ -0,0 +1,23 @@
+      } else {
+        logger.error(`failed to publish update for ${profileUpdate.userId}`);
+        return Promise.reject(
+          `search service returned status code ${res.status}`
```
Same here.

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+  handler() {
+    return (req, res) => {
+      this.publishToCIS(req.body)
+        .then(x => this.pollUpdateId(x))
```
Let's replace x with something more descriptive.

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+    return (req, res) => {
+      this.publishToCIS(req.body)
+        .then(x => this.pollUpdateId(x))
+        .then(x => this.getProfileByUserId(x))
```
Same here.

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+      this.publishToCIS(req.body)
+        .then(x => this.pollUpdateId(x))
+        .then(x => this.getProfileByUserId(x))
+        .then(x => res.json(x))
```
Same here.

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+        .then(x => res.json(x))
+        .catch(e => {
+          logger.error(`failed to update profile: ${e}`);
+          res.status(503).json({});
```
Should we return some sort of message here?

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+      if (e.response) {
+        throw new CallError(
+          `error retrieving update status for ${updateId}\
+(status code: ${e.response.status})`,
```
Is that supposed to look that way ?

fiji-flo replied:
Yeah, I can change to look like the next one below. And I will.

Multiline stings…

johngian commented on lib/userUpdate.js:
```diff
@@ -0,0 +1,93 @@
+      throw new CallError(
+        `publishing profile for ${
+          profileUpdate.userId
+        } faild with empty response`
```
``faild``

johngian replied:
nit: Can we wrap this to a single line?

johngian commented on test/configTest.js:
```diff
@@ -0,0 +1,70 @@
+chai.use(chaiAsPromised);
+chai.should();
+
+describe("Everything configy", () => {
```
``configy``

johngian commented on test/configs.js:
```diff
@@ -0,0 +1,13 @@
+const EMPTY = {
```
Why ``EMPTY`` and not something more descriptive, is that required for some for reason ?

Gregoor commented on lib/config.js:
```diff
@@ -0,0 +1,69 @@
+  try {
+    const u = new URL(s);
+  } catch (e) {
+    throw new Error(`should be an vaild URL (${e})`);
```
\* a valid

Gregoor commented on lib/config.js:
```diff
@@ -0,0 +1,69 @@
+  }
+  if (error) {
+    throw new Error(`Config file malformed: ${error}!`);
+  }
```
Might be nice to get all errors at once:
```javascript
const errors = VALID.map(([key, validate]) => {
  if (!(key in config)) {
    return `missing ${key}`;
  }
  try {
    validate(config[key]);
  } catch (e) {
    return `${key} ${e.message}`;
  }
  return '';
});
if (errors.length > 0) {
  throw new Error(`Config file malformed: \n${errors.join('\n')}!`);
}
```

Gregoor commented on lib/externalUpdate.js:
```diff
@@ -0,0 +1,51 @@
+  handler() {
+    return (req, res) => {
+      ExternalUpdate.getUserId(req.body)
+        .then(userId => this.getAndDistributeUpdate(userId))
```
Out of curiosity: Why no async/await?

fiji-flo replied:
I think this is a nice chain and `Promise.then` looks nicer here.

Gregoor commented on lib/externalUpdate.js:
```diff
@@ -0,0 +1,51 @@
+  async getAndDistributeUpdate(userId) {
+    return this.personApi(userId)
+      .then(updatedProfile => this.searchService(updatedProfile))
+      .then(updatedProfile => this.orgchartService(updatedProfile));
```
Are those promises run sequentially because the orgchart should only update if the search update succeeded? Or could they also run in parallel?

fiji-flo committed:
Adopting PR feedback.

- tidy up \*ignore files
- rename test config
- typos
- formatting
- error messaging

fiji-flo committed:
Use convict for config parsing.

fiji-flo committed:
Fix tests for config.

fiji-flo committed:
Some cleanup and typo fixes.
